proxy/Main.cc

```diff
@@ -1934,7 +1951,30 @@ main(int /* argc ATS_UNUSED */, const char **argv)
       if (delay_p && ink_atomic_cas(&delay_listen_for_cache_p, 0, 1)) {
         Debug("http_listen", "Delaying listen, waiting for cache initialization");
       } else {
-        start_HttpProxyServer(); // PORTS_READY_HOOK called from in here
+        // The problem here is we assume the et_net threads has started, but during
+        // some testing that is not always the case, some times start_HttpProxyServer
+        // will be called before the et_net threads finish initializing, which causes
+        // crashes due to uninitilized pointers being NULL. This happens only when
+        // we set the accept_threads to be 0 in the config, since this means the
+        // accept_threads is attached to the et_net threads. The first check here takes
+        // care of this special case of accept_threads being set to 0.
+        //
+        // The second check here is to make sure that ALL et_net threads has finished
+        // their initilization, so there are no uninitialied variables when we start
+        // the accept_threads. We will try to change wait_for_et_net_ready from 0 to 1,
+        // if it succeeded it means that this is triggered before other places in this
+        // race condition, so we shouldn't start the accept_threads here. However if it
+        // did change wait_for_et_net_ready successfully, it means we can start the
+        // accept_threads here since other places in the race condition has finished
+        // their job.
+        if (num_accept_threads == 0) {
+          int expected = 0;
+          if (!wait_for_et_net_ready.compare_exchange_strong(expected, 1)) {
+            start_HttpProxyServer();
```

Comment:
It might be cleaner to use a mutex and unlock it here and have the et_net thread call start_HttpProxyServer() all the time.